lib/instrument/transformer/nskeyed.js

```diff
@@ -0,0 +1,431 @@
+  decodeArchive (archive) {
+    return archive.decode('capabilities-dictionary');
+  }
+}
+
+class NSNull extends BaseArchiveHandler {
+  decodeArchive () {
+    return null;
+  }
+}
+
+/**
+ * decode and encode Archive of currently known data formats
+ */
+const UNARCHIVE_CLASS_MAP = {
+  DTTapMessagePlist,
+  DTSysmonTapMessage: DTTapMessagePlist,
+  DTTapHeartbeatMessage: DTTapMessagePlist,
+  DTTapMessageArchive: DTTapMessagePlist,
+  DTKTraceTapMessage: DTTapMessagePlist,
+  ErrorArchive: NSError,
+  ExceptionArchive: NSException,
+  NSDictionary: DictArchive,
+  NSMutableDictionary: DictArchive,
+  NSArray: ListArchive,
+  NSMutableArray: ListArchive,
+  NSMutableSet: ListArchive,
+  NSSet: ListArchive,
+  NSDate,
+  NSError,
+  NSException,
+  NSMutableString,
+  NSMutableData,
+  NSNull,
+  NSUUID,
+  NSURL,
+  XCTCapabilities
+};
+
+/**
+ * Capable of unpacking an archived object tree in the NSKeyedArchive format.
+ * Apple's implementation can be found here:
+ * https://github.com/apple/swift-corelibs-foundation/blob/main/Sources/Foundation/NSKeyedUnarchiver.swift
+ */
+class Unarchive {
+  constructor (inputBytes) {
+    this.input = inputBytes;
+    this.unpackedUids = {};
+    this.topUID = NULL_UID;
+    this.objects = [];
+  }
+
+  unpackArchiveHeader () {
+    const plist = plistlib.parseBuffer(this.input)[0];
+    if (plist.$archiver !== NSKEYEDARCHIVER) {
+      throw new Error(`unsupported encoder: ${plist.$archiver}`);
+    }
+    if (plist.$version !== NSKeyedArchiveVersion) {
+      throw new Error(`expected ${NSKeyedArchiveVersion}, got ${plist.$version }`);
+    }
+    const top = plist.$top;
+    const topUID = top.root;
+    if (!topUID) {
+      throw new Error(`top object did not have a UID! dump: ${JSON.stringify(top)}`);
+    }
+    this.topUID = topUID;
+    this.objects = plist.$objects;
+  }
+
+  /**
+   * use the UNARCHIVE_CLASS_MAP to find the unarchiving delegate of a uid
+   */
+  classForUid (index) {
+    const meta = this.objects[index.UID];
+    const name = meta.$classname;
+    const klass = UNARCHIVE_CLASS_MAP[name];
+    if (!klass) {
+      throw new Error(`no mapping for ${name} in UNARCHIVE_CLASS_MAP`);
+    }
+    return klass;
+  }
+
+  decodeKey (obj, key) {
+    const val = obj[key];
+    return _.isNil(val?.UID) ? val : this.decodeObject(val);
+  }
+
+  decodeObject (index) {
+    if (index === NULL_UID) {
+      return null;
+    }
+    const obj = this.unpackedUids[index];
+    if (obj === CYCLE_TOKEN) {
+      throw new Error(`archive has a cycle with ${index}`);
+    }
+    if (!_.isUndefined(obj)) {
+      return obj;
+    }
+    const rawObj = this.objects[index.UID];
+    this.unpackedUids[index.UID] = CYCLE_TOKEN;
+
+    if (!rawObj?.$class) {
+      this.unpackedUids[index.UID] = obj;
+      return rawObj;
+    }
+    const klass = this.classForUid(rawObj.$class);
+    const klassObj = new klass().decodeArchive(new ArchivedObject(rawObj, this));
```

Comment:
some classes, like NSUUID require constructor arguments to be provided. How do we handle this situation here?

Reply:
We do not need constructor arguments in the decoding process. constructor arguments required only when encoding process
Refer to nskeyed-specs.js unit test case

Reply:
Could you please help me to find the place where arguments are passed to the corresponding constructors for further encoding?

Reply:
The instance has been created before encode . Like this
``` javascript
const data = {'NSURL': new NSURL(null, 'file://PlugIns/WebDriverAgentRunner.xctest')};
const archiveData = archive(data);
```
After that NSURL will be encoded in this code. `obj` is an instance of the NSURL
```javascript
if (objName in UNARCHIVE_CLASS_MAP) {
  archiveObj.$class = this.uidForArchiver(objName);
  obj.encodeArchive(obj, new ArchivingObject(archiveObj, this));
```